### PR TITLE
Meaningful `toString`, unit tests, `<id>` vs `<text>` differentiation, and more

### PR DESCRIPTION
**Pull request by tglaeser**
I'm new to Kotlin, so it is very likely that some of the proposed changes don't completely benefit from the language, I'm looking for guidance from you here.

I started this with the thought that the token literals rather belong to their corresponding `IDLType` instead of being fields in `IDLGrammar`; this would then also allow to have a meaningful `toString` method for each type.

In addition I wanted to make sure that all unit tests cover the `IDLType` to Candid string to `IDLType` round trip; for this I added the needed dependencies to the test source set to use my favorite testing framework Spock. As a start I added coverage for all the use cases I found in class `GrammarTest`. In a next step one then could add test cases that would in fact test the generated Kotlin code.

Other than that I tried to stay away from the actual implementation as much as I could, but ran into an issue with respect to the following part of the Candid grammar:
```
<name> ::= <id> | <text>
<id>   ::= (A..Z|a..z|_)(A..Z|a..z|_|0..9)*
<text> ::= "<char>*"
```
The proposed changes should reflect the above grammar more closely.

In this context I renamed some fields, e.g. `pActorTypeType`, to get a better grasp on the differentiation between `<actortype>` and `<reftype>`.

In the end this PR became larger then initially expected, hope that's not a problem.

Please advise how to proceed, I'm interested in contributing more with the goal of getting the **Cons** section from the readme eliminated.

### CODE REVIEW

**Comment by seniorjoinu** [comment]
Hello there! What a great job you did here! 

I apologize for such a long time without any reaction to this PR. If I go unreachable again - text me in telegram (https://t.me/joinu14)

Let me go step by step for your message:

> I started this with the thought that the token literals rather belong to their corresponding IDLType instead of being fields in IDLGrammar; this would then also allow to have a meaningful toString method for each type.

That's a great idea. This is an awesome refactor.

> In addition I wanted to make sure that all unit tests cover the IDLType to Candid string to IDLType round trip; for this I added the needed dependencies to the test source set to use my favorite testing framework Spock. As a start I added coverage for all the use cases I found in class GrammarTest

This is also a nice thing to have. The only concern I have is that you used Groovy instead of Kotlin, but for now it is fine, if it runs without any additional steps. Thank you!

> Other than that I tried to stay away from the actual implementation as much as I could, but ran into an issue with respect to the following part of the Candid grammar

I gonna check it out right now. It seems like you erased the sealed inheritance part from these guys. Gonna text you an update for this one later.

> In this context I renamed some fields, e.g. pActorTypeType, to get a better grasp on the differentiation between <actortype> and <reftype>.

This is also good. No questions on that.

> Please advise how to proceed, I'm interested in contributing more with the goal of getting the Cons section from the readme eliminated.

Once again, I really appreciate your contribution to this project. Good job!
If you wish to contribute more, you can check out the `Issues` section of this repo - there are already some tasks and many of them are good for understanding the internal mechanics of this library - just comment one of them without an assignee "I would like to do that" and I will assign it to you. And also feel free to ask anything about issues - some of them might not have a comprehensive description.

As for this PR, I'm gonna check if everything works fine and merge it. If there is something not okay, I'll add some comments to your code.

Have a nice day.

**Comment by tglaeser** [comment]
Thank you Sir - I also enjoyed reading your code.

I also was hesitant with adding a dependency on Groovy, but believe it should be fine as long as it is restricted to the `test` configurations only; of course it would be great to have a Spock for Kotlin port available, but this might not be possible at all as Spock heavily depends on Groovy's dynamic language features; ultimately I believe the advantage with Groovy for testing is that it results in more concise/readable test cases compared to writing them in Java or even Kotlin.

I'm planing to come back to this project but I'm currently working on a PR for the sibling `candid-kt-gradle-plugin` project, I believe it should operate on a `candid` source directory set, stay tuned...

**Comment by seniorjoinu** [comment]
Is there a problem with JUnit? For my entire career it works just fine with Kotlin.

> I believe it should operate on a candid source directory set, stay tuned

Feel free to improve that as well, but keep it flexible - it should be possible to have either single project for kotlin-code and IC-code or have different projects for each of them.

**Comment by tglaeser** [comment]
> Is there a problem with JUnit?

Well, first there is conciseness, then JUnit is not a full fledged testing framework as it lacks support for mocking and stubbing, so eventually one would have to add dependencies to modules like Mockito and PowerMock which then leads to even less conciseness.

So you prefer to keep the language in this project 100% Kotlin? As of now mocking/stubbing is not used yet, so it should be easy to replace all the plain boolean expressions with the JUnit assertion API; let me know.

**Comment by seniorjoinu** [comment]
I'm just afraid. Like one day we'll need an update for these tests and there will be no one who knows Groovy around (I don't know it). Like the project is called "candid-kt" and it will attract mostly Kotlin developers.

I'm agree that JUnit sometimes needs a bit of tuning.

I think we can do something like this: for now let it be as it is, but you should somehow redo these tests in Kotlin. Maybe dig a little and find some testing framework that will be better than JUnit (there is a lot of innovation in the Kotlin community - you can find almost anything).
What do you think?

**Comment by tglaeser** [comment]
Sounds like a plan, I will do that.

Having said that, any seasoned developer who can read and write Java/Kotlin can read and write a Spock specification; I would not want to use Groovy for anything else either.

BTW, I already posted this question to the Spock community and got the expected answer; dynamic language features are probably essential for the mocking/stubbing.

I pledge to replace the Spock specifications I introduced with any mocking/stubbing framework for the Kotlin language as soon as it becomes available; let's continue this discussion over another channel...

File: src/main/kotlin/senior/joinu/candid/IDLType.kt [hunk 1]
```diff
@@ -235,150 +249,147 @@ sealed class IDLType {
...
             override fun poetize() = CodeBlock.of("%T($opcode)", Custom::class).toString()
         }
 
         data class Future(val opcode: Int) : Other() {
+            val text: String = TODO("Not yet implemented")
+            override fun toString() = TODO("Not yet implemented")
             override fun poetize() = CodeBlock.of("%T($opcode)", Future::class).toString()
         }
     }
+
```
Comment (seniorjoinu) [inline src/main/kotlin/senior/joinu/candid/IDLType.kt]: Please, consider moving this function in the `Utils.kt` file.

Also, in Kotlin you can use """text""" (3 pairs of double-quotes) to produce multiline strings (all whitespace characters are preserved inside these). UPD: this part is not required to do

Comment (seniorjoinu) [reply src/main/kotlin/senior/joinu/candid/IDLType.kt]: And could you also make it public?

File: src/main/kotlin/senior/joinu/candid/IDLParser.kt [hunk 2]
```diff
@@ -287,19 +281,22 @@ object IDLGrammar : Grammar<IDLProgram>() {
...
+            companion object { const val pattern: String = "0x[0-9a-fA-F][_0-9a-fA-F]*" }
+        }
     }
 
-    class TextVal(override val value: String) : IDLToken(),
-        IDLTextToken
-
-    class Name(override val value: String) : IDLToken(),
-        IDLTextToken
+    data class TextVal(override val value: String) : IDLName {
```
Comment (seniorjoinu) [inline src/main/kotlin/senior/joinu/candid/IDLParser.kt]: Please make it also inherit the `IDLToken()` parent.

Comment (seniorjoinu) [reply src/main/kotlin/senior/joinu/candid/IDLParser.kt]: It is needed to make `when` expression work correctly.

**Review by seniorjoinu — CHANGES_REQUESTED** [review]
Nicely done!
Tests pass on my machine.

Fix these little guys and I'll merge.
Thank you.

**Comment by seniorjoinu** [comment]
> Sounds like a plan, I will do that.

That's awesome. I'll create an issue for that and assign it to you.
Thank you.

**Comment by seniorjoinu** [comment]
For some reason I can't make you an assignee. Seems like you should do it yourself.

File: src/main/kotlin/senior/joinu/candid/IDLParser.kt (hunk 2 above)
Comment (tglaeser) [reply src/main/kotlin/senior/joinu/candid/IDLParser.kt]: Done.

File: src/main/kotlin/senior/joinu/candid/IDLType.kt (hunk 1 above)
Comment (tglaeser) [reply src/main/kotlin/senior/joinu/candid/IDLType.kt]: Did that; however moving the function under Kotlin package `transpile` seems a little misleading; also I couldn't find a more readable representation using the triple-double-quote notation as one has to iterate over the items and indent and append a new line character to each item; please improve upon as you like.

**Review by seniorjoinu — APPROVED** [review]

**Pull request merged**